### ISSUE TITLE
Status update on paperclip?

### ISSUE THREAD

**Issue by mokolabs** [thoughtbot/paperclip#2597]
It's been almost a week since you guys closed all outstanding issues.

While I can understand your desire to deprecate this gem in favor of ActiveStorage, you owe the community a lengthy explanation (and plan for going forward) because paperclip is the primary attachment gem for tens of thousands of production Rails apps.

FWIW, I'm hoping you give the paperclip community an option to move forward without you. Sure, ActiveStorage looks great, but we need to support apps that currently rely on paperclip until ActiveStorage fully matures and offers a clean migration path for existing paperclip users -- even if it means doing so without thoughtbot's help.

## Deprecation notice

Paperclip is currently undergoing
[deprecation in favor of ActiveStorage](https://github.com/thoughtbot/paperclip/blob/master/MIGRATING.md).
Maintainers of this repository will no longer be tending to new issues. We're leaving the issues page
open so Paperclip users can still see & search through old issues, and continue existing discussions
if they wish.

**Comment by sidraval** [comment]
Hi @mokolabs ,

What we mean to do by deprecating the gem is to tell the community that they should no longer expect thoughtbot to support Paperclip. We understand that it’s not realistic to expect people to migrate to a new solution overnight. However, deprecating the gem doesn't mean that people have to stop using it. After merging the remaining PRs, we expect Paperclip to be in a reasonable place to give the teams using it many weeks, months, or perhaps longer to make their decisions.

Paperclip has been languishing for a while now, as you can see in the commit history over the last year or more. We at thoughtbot feel this deprecation is largely a reflection of how Paperclip has been maintained in recent history.

The community has every option to move forward without thoughtbot. The [license](https://github.com/thoughtbot/paperclip/blob/master/LICENSE) is permissive.

**Comment by benbjohnson** [comment]
> _"you owe the community a lengthy explanation"_

@mokolabs This is not appropriate to post to an open source project that you have paid no money for and that Thoughtbot has never committed to any long term support. They don't owe you or anyone else anything. The community owes them gratitude for developing the project and being stewards of it for so many years.

@sidraval Thanks for all the hard work. 🚀

**Comment by mokolabs** [comment]
@benbjohnson I'm pretty familiar with how open source works.

I'm also deeply thankful for thoughtbot's tremendous contribution to the Rails community by creating Paperclip and maintaining it all of these years.

And I also understand thoughtbot's decision to deprecate Paperclip, and I totally respect it.

...

My point, however, was that thoughtbot can do a better job announcing the deprecation and easing the migration pain for the Paperclip community. (And I should note that since I opened this issue, @sidraval has responded and clarified a few things... and that's a good start. Thanks, @sidraval!).

But, ask yourself, if you were maintaining another open source project with a large user base, what would be the best way to communicate a big change to a project like this? Do you think just turning off issues and adding a deprecation notice is all you need to do to shut down a project?

If, say, later today, DHH announced Rails is deprecated _in a GitHub issue_ and suggested we all migrate to Ember, would that be a good way to deprecate Rails? I don't think so.

I originally added this issue because I noticed that thoughtbot -- based on its own statements in the tweets below -- was planning to give the Paperclip community a more complete explanation and plan for the deprecation, and had yet to do so.

<img width="629" alt="screen shot 2018-05-07 at 8 32 21 am" src="https://user-images.githubusercontent.com/444/39710351-710accaa-51d1-11e8-8aa9-b2e9c934be7e.png">

And, yes, in open source, no one technically owes anyone anything. But, in general, I hold the Ruby and Rails community to a higher standard and that includes thoughtbot -- most especially because, in this case, the deprecation of Paperclip impacts thousands of developers with billons of assets sitting on Amazon S3 who will need to spend **substantial time and money** on migrating away from paperclip. It's completely non-trivial.

So how can thoughtbot do a better job with this?

Here are a few things I would recommend:

- Publish a blog post to the thoughbot blog officially announcing the deprecation and suggesting ways to migrate to Shrine and ActiveStorage.

- Send out a tweet with a link to the blog post

- Find future maintainers from the paperclip community to take over the project and migrate it to a new home (but keep the existing commit and issue history intact). [And, yes, the license is permissive, and someone could just clone it, but that's not optimal.]

- Update the project README with a short description of the deprecation and a link to the blog post

- Add a deprecation notice to https://thoughtbot.com/services/ruby-on-rails

- Notify Ruby gem directories that paperclip is now deprecated

- Update previous thoughtbot blog posts on paperclip with a mini deprecation notice

See the difference? 😃

Lastly, while I don't see myself as a future paperclip maintainer, if thoughtbot needs help working on any of these suggestions, I would be happy to contribute. I'm a solid technical writer, so I could help with a draft of the blog post and updates to the README.

Also, I'm a frequent contributor to [Shrine](https://github.com/shrinerb/shrine) and [image_processing](https://github.com/janko-m/image_processing), so I'm going to take a pass at updating docs to make it super clear how folks can migrate from Paperclip.

Let me know how I can help, @sidraval!

**Comment by sidraval** [comment]
@mokolabs Indeed, closing the issues was not a sensible way to proceed, which is clear to us now 😄 

We are planning on, minimally:
* Writing a blog post & a corresponding tweet
* Updating the README

Possibly more, but those two things are the priority after we get as many of the remaining PRs merged as we can.

Updating the Shrine docs would be great 👌 It looks like CarrierWave already has a way to migrate in their documentation, and we're working on bettering our guide to migrating to Active Storage.

**Comment by SleeplessByte** [comment]
For what it's worth, @mokolabs cc @sidraval , @janko-m has already written https://github.com/shrinerb/shrine/blob/master/doc/paperclip.md a while back.

**Comment by mokolabs** [comment]
I'm glad to hear it, @sidraval. Let me know if I can be of service. 😄

**Comment by hermanomark** [comment]
goodbye paperclip :(

**Comment by maxence33** [comment]
Just one question, does Paperclip work with latest Ruby and Rails version as of now ? I may upgrade to Rails 5.2 in a near future (I am already with Ruby 2.5) but I have not planned migrations and code change for ActiveStorage yet..

**Comment by mokolabs** [comment]
Yes, Paperclip still works with Rails 5.2, @maxence33.

**Comment by collimarco** [comment]
Paperclip is better than ActiveStorage in many situations. I have used Paperclip for many years and only a few weeks ago I started a new project with Paperclip (after a careful comparison with ActiveStorage). I find that ActiveStorage is too much opinionated, cumbersome and it only fits some applications, not all.

I hope that Paperclip will find a new maintainer.
